src/main/java/org/apache/sling/feature/io/artifacts/spi/ArtifactProvider.java

```diff
@@ -43,8 +44,20 @@
 
     /**
      * Shutdown the provider.
+     * @deprecated Use {@link #close()} instead.
      */
-    void shutdown();
+    @Deprecated
+    default void shutdown() {
+        
```

Comment:
Why is this switching to a default implementation?

Reply:
New implementation should only implement close(). Therefore the default impl of the deprecated method.

Reply:
But as close is also a default method (which makes sense for compatibility) this is kind of a strange situation and can easily lead to that none of them is implemented.
I have the feeling we should not deprecate shutdown but just add close calling shutdown to suppot Closeable

Reply:
The idea is that provider implementations rather implement `close()` instead of `shutdown()` to be able to propagate IOExceptions. Having two different methods for shutting down and only the suboptimal signature (`shutdown()`) as mandatory implementation feels wrong. With regards to no shutdown/close method being mandatory I think this totally fine, as not every artifact provider necessarily needs to do clean up.

Reply:
I'm not sure - shutdown not throwing a checked exception is intentional - what should a client do if shutdown is failing? I think shutdown is different from closing an output stream where close is required to fully work to not loose data. If shutdown of the provider fails - what damage does it do?

Reply:
The ArtifactManager is treating IOExceptions differently than other unchecked exceptions (https://github.com/apache/sling-org-apache-sling-feature/pull/25/files#diff-36fc363deea378f4f1f538a997cebb80f6aca64974b2aef836f91371c44e06ecR114 and https://github.com/apache/sling-org-apache-sling-feature/pull/25/files#diff-36fc363deea378f4f1f538a997cebb80f6aca64974b2aef836f91371c44e06ecR143) to not hide the original exception/still close the other providers. Also since all File based operations might throw IOE in Java otherwise the provider would need to wrap it in a RuntimeException (if implementing shutdown).
In certain cases it might make sense to retry calling close() as some IOExceptions are transient....

Reply:
But shouldn't such a retry logic be part of shutdown? It seems to be a high burden to clients to figure out whether the IOException is transient or persistent.
Right now, both provider and manager have a shutdown method not throwing a checked exception. It is expected that the implementation does whatever is required to shutdown. With introducing close() to both, the provider and the manager, you create this new situation where the provider now can throw on close - which then needs to be handled by the manager.
Therefore my expectation is that shutdown never throws (a checked exception), we can implement Closeable as a convenience but close() will never throw an IOException.

Reply:
> With introducing close() to both, the provider and the manager, you create this new situation where the provider now can throw on close - which then needs to be handled by the manager.

AM had a close() method before this PR. So it may have thrown IOExceptions already there. The only change right now is that providers (instead of just logging with their own logging framework) propagate their exception to the manager. Although this doesn't matter that much if you only think about the Default provider, it may make a difference for 3rd party providers. To log all those places centrally, the exceptions should hit the manager.

Reply:
No, AM implements close without throwing an exception, see https://github.com/apache/sling-org-apache-sling-feature/blob/master/src/main/java/org/apache/sling/feature/io/artifacts/ArtifactManager.java#L126
AM also implements AutoCloseable - so AM can already be used in a try-with-resources block. Nothing needs to change there, and we can apply the same pattern to the ArtifactProvider - let it implement AutoCloseable with a default method close() not throwing an exception and calling shutdown

Reply:
All you said is true, but my argument was "To log all those places centrally, the exceptions should hit the manager.". Do you disagree with that?

Reply:
Yes I disagree :) using exceptions just to ensure central logging sounds wrong to me